File: kotlin/pig-latin/src/main/kotlin/PigLatin.kt
object PigLatin {
    fun translate(phrase: String): String {
        return phrase.split(" ").fold("") { acc, s ->
            acc.plus(checkWord(s)).plus(" ")
        }.trim()
    }

    fun checkWord(word: String): String {
        return when {
            word.checkRuleFive() -> word.applyRuleFive()
            word.checkRuleThree() -> word.applyRuleThree()
            word.checkRuleOne() -> word.applyRuleOne()
            word.checkRuleTwo() -> word.applyRuleTwo()
            word.checkRuleFour() -> word.applyRuleFour()
            else -> word
        }
    }

    fun String.checkRuleOne(): Boolean {
        val prefix = listOf("a", "e", "i", "o", "u", "xr", "yt").firstOrNull() {
            this.startsWith(it, true)
        }
        return prefix != null
    }

    fun String.applyRuleOne() : String {
        return this + "ay"
    }

    fun String.checkRuleTwo(): Boolean {
        return this.indexOfFirst { listOf('a', 'e', 'i', 'o', 'u').contains(it.lowercaseChar()) } > 0
    }

    fun String.applyRuleTwo(): String {
        val index = this.indexOfFirst { listOf('a', 'e', 'i', 'o', 'u').contains(it.lowercaseChar()) }
        return this.substring(index).plus(this.take(index)).plus("ay")
    }

    fun String.checkRuleThree(): Boolean {
        return Regex("""^[^aeiou](qu).*""").matches(this)
    }

    fun String.applyRuleThree(): String {
        val index = if (this.first() == 'q') 2 else 3
        return this.substring(index).plus(this.take(index)).plus("ay")
    }

    fun String.checkRuleFour(): Boolean {
        return Regex(""".y""").matches(this) ||
                Regex("""[^aeiou]+y.*""").matches(this)
    }

    fun String.applyRuleFour(): String {
        val index = this.indexOf('y', ignoreCase = true)
        return this.substring(index).plus(this.take(index)).plus("ay")
    }

    fun String.checkRuleFive(): Boolean {
        return Regex("""^(qu).*""").matches(this)
    }

    fun String.applyRuleFive(): String {
        return this.substring(2).plus(this.take(2)).plus("ay")
    }
}

File: python/pig-latin/pig_latin.py
def translate(text: str):
    string = ""
    return (string.join(translate_word(word)) for word in text.split(' '))


def translate_word(text):
    if passes_rule_three(text):
        return apply_rule_three(text)
    if passes_rule_one(text):
        return apply_rule_one(text)
    if passes_rule_two(text):
        return apply_rule_two(text)


def is_vowel(char):
    return char in ('a', 'e', 'i', 'o', 'u')


def is_vowel_or_y(char):
    return char in ('a', 'e', 'i', 'o', 'u', 'y')


def passes_rule_one(text):
    return is_vowel(text[0]) or text[:2] in ('xr', 'yt')


def apply_rule_one(text):
    return text + 'ay'


def passes_rule_two(text):
    return not is_vowel(text[0])


def apply_rule_two(text):
    first_vowel_index = next(i for i, v in enumerate(text) if is_vowel_or_y(v))
    return str[first_vowel_index:] + str[0:first_vowel_index] + 'ay'


def passes_rule_three(text):
    return text[0:2] == 'qu' or (not is_vowel(text[0]) and text[1:3] == 'qu')


def apply_rule_three(text):
    if text[0] == 'q':
        return text[2:] + text[:2] + 'ay'
    return text[3:] + text[:3] + 'ay'
